source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,598 @@
+    const ir::Instruction* type) const {
+  assert(type->opcode() == SpvOpTypeVector ||
+         type->opcode() == SpvOpTypeMatrix);
+  const ir::Operand& op = type->GetInOperand(1u);
+  assert(op.words.size() <= 2);
+  size_t len = 0;
+  for (uint32_t i = 0; i != op.words.size(); ++i) {
+    len |= (op.words[i] << (32 * i));
+  }
+  return len;
+}
+
+ir::Instruction* ScalarReplacementPass::GetStorageType(
+    const ir::Instruction* inst) const {
+  assert(inst->opcode() == SpvOpVariable);
+
+  uint32_t ptrTypeId = inst->type_id();
+  uint32_t typeId =
+      get_def_use_mgr()->GetDef(ptrTypeId)->GetSingleWordInOperand(1u);
+  return get_def_use_mgr()->GetDef(typeId);
+}
+
+bool ScalarReplacementPass::CanReplaceVariable(
+    const ir::Instruction* varInst) const {
+  assert(varInst->opcode() == SpvOpVariable);
+
+  // Can only replace function scope variables.
+  if (varInst->GetSingleWordInOperand(0u) != SpvStorageClassFunction)
+    return false;
+
+  const ir::Instruction* typeInst = GetStorageType(varInst);
+  if (!CheckType(typeInst)) return false;
+
+  if (!CheckUses(varInst)) return false;
+
+  return true;
+}
+
+bool ScalarReplacementPass::CheckType(const ir::Instruction* typeInst) const {
+  switch (typeInst->opcode()) {
+    case SpvOpTypeStruct:
+      // Don't bother with empty structs or very large structs.
+      if (typeInst->NumInOperands() == 0 ||
+          typeInst->NumInOperands() > MAX_NUM_ELEMENTS)
+        return false;
+      return true;
+    case SpvOpTypeArray:
+      if (GetArrayLength(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+    // Specifically including matrix and vector in an attempt to reduce the
+    // number of vector registers required.
+    case SpvOpTypeMatrix:
+    case SpvOpTypeVector:
+      if (GetNumElements(typeInst) > MAX_NUM_ELEMENTS) return false;
+      return true;
+
+    case SpvOpTypeRuntimeArray:
+    default:
+      return false;
+  }
+}
+
+bool ScalarReplacementPass::CheckUses(const ir::Instruction* inst) const {
+  bool ok = true;
+  VariableStats stats = {0, 0};
+  CheckUses(inst, &stats, &ok);
+
+  // TODO(alanbaker): Extend this to some meaningful heuristics about when SRoA
+  // is valuable.
+  if (stats.num_partial_accesses == 0) ok = false;
+
+  return ok;
+}
+
+void ScalarReplacementPass::CheckUses(const ir::Instruction* inst,
+                                      VariableStats* stats, bool* ok) const {
+  get_def_use_mgr()->ForEachUse(
+      inst, [this, stats, ok](const ir::Instruction* user, uint32_t index) {
+        switch (user->opcode()) {
+          case SpvOpAccessChain:
+          case SpvOpInBoundsAccessChain:
+            if (index == 2u) {
+              uint32_t id = user->GetSingleWordOperand(3u);
+              const ir::Instruction* opInst = get_def_use_mgr()->GetDef(id);
+              if (!ir::IsCompileTimeConstantInst(opInst->opcode())) {
+                *ok = false;
+              } else {
+                CheckUsesRelaxed(user, ok);
```

Comment:
Not obvious  what you are looking for and why.  Can you add to the comments in the header file?  My guess is that you want to make sure the address does not "escape".  However, why would if matter if the address of the member escapes as say a parameter on a function call?  You can pass it the address of the new variable.

Reply:
This checks that the pointer ends up in a load or store and not something else. If this were a kernel, I could do ptrtoint and then weird stuff. I don't want that to be allowed.